[WIP] P5-212: Continous Integration (#131)
aau-hcl-p5/main: report/6.Conclusion/development_practices.tex

Slack allows the use of bots that can be triggered by events in Jira, Github or TeamCity which were used for issue tracking, version control and continuous integration, respectively.
This centralized all notifications regarding the project on one platform.

\section{Issue tracking}
Jira is an issue-tracking system with a large variety of features and integrations, that allows creating kanban boards.
Having an online issue-reporting system made it easy for the group members to add, update and read issues.
During the last two sprints, a physical kanban board was used instead, as it allowed for easier tracking of progress.

\section{Version Control}
For version control the Git protocol, through GitHub, was used, to keep track of different versions of the project, 2 main branches was used: a master branch, a develop branch and a series of feature branches.
Feature branches were used for solving issues, meaning that each issue on the kanban board, had a corresponding feature branch.

The develop branch served as sprint branches, and all feature branches are merged into the develop branch during the week.

> These merges could only happen with a pull-request that had been reviewed and approved by one or more members, and had successfully passed all tests.

Reviewer: ```suggestion
These merges could only happen with a pull request that had been reviewed and approved by one or more members, and had successfully passed all tests.
```

Conclusively, the development was greatly improved by the iterative practices, both regarding the abstractive practices of development, as well as the agile principles followed through sprints, testing and continuous integration.
The practices improves cooperation as well as velocity, especially during the final weeks of development, and should be a good foundation for a new project in the coming semesters.

\section{Continous Integration}

> For Continous Integration, the group used TeamCity with custom build scripts, to ensure that all builds were built correctly\cite{TeamCityHomepage}.

Reviewer: ```suggestion
For continuous integration, the group used TeamCity with custom build scripts, to ensure that all builds were built correctly\cite{TeamCityHomepage}.
```

Conclusively, the development was greatly improved by the iterative practices, both regarding the abstractive practices of development, as well as the agile principles followed through sprints, testing and continuous integration.
The practices improves cooperation as well as velocity, especially during the final weeks of development, and should be a good foundation for a new project in the coming semesters.

\section{Continous Integration}

> The group has utilised a Continous Integration tool, called TeamCity, with custom build scripts to ensure all builds and their integrity\cite{TeamCityHomepage}.

Reviewer: ```suggestion
For Continous Integration, the group used TeamCity with custom build scripts, to ensure that all builds were built correctly\cite{TeamCityHomepage}.
```

Reviewer 2: Skal Continous Integration ikke stå med småt?

> To avoid a faulty commit and merge with a erroneous build, the TeamCity server provided GitHub pull-requests with status updates via the GitHub API, to ensure that failing builds/branches were not mergeable.

Reviewer 3: ```suggestion
To avoid a faulty commit being merged resulting in an erroneous build, the TeamCity server provides GitHub pull-requests with status updates via the GitHub API, to ensure that failing builds/branches were not mergeable. 
```

> To avoid a faulty commit being merged resulting in an erroneous build, the TeamCity server provides GitHub pull-requests with status updates via the GitHub API, to ensure that failing builds/branches were not mergeable. 

Reviewer: ```suggestion
To prevent a branch with a failed build from being merged, the TeamCity server provides GitHub pull-requests with status updates via the GitHub API, to ensure that failing branches were not mergeable. 
```

Reviewer 2: Må ikke lave en suggestion af Lisbys suggestion - men der skal ikke være bindestreg i pull requests

> Furthermore, the TeamCity server builds are producing artifacts, such as code-coverage reports and a PDF-document of this report is produced per build, to ease and maintain a centralised build-platform, that is independent of developer environments.

Reviewer 3: ```suggestion
Furthermore, the TeamCity server builds are producing artifacts, such as code-coverage reports and a PDF-document of this report is produced as a result of the build, to ease and maintain a centralised build-platform, that is independent of developer environments.
```

> Furthermore, the TeamCity server builds are producing artifacts, such as code-coverage reports and a PDF-document of this report is produced as a result of the build, to ease and maintain a centralised build-platform, that is independent of developer environments.

Reviewer: ```suggestion
Furthermore, the TeamCity server builds produces artifacts, such as code-coverage reports and a PDF-document of the report.
Thus an developer environments independent version of the project is kept, meaning that some developer environment specific errors are avoided.
```

Reviewer: Ved ikke helt med det sidste jeg har forslået her, men der er lidt rod i den sætning lige meget hvad.
